zerver/tests/test_message_send.py

```diff
@@ -1325,6 +1325,104 @@ def test_message_bot_mentions(self) -> None:
         self.assertEqual(user_message.message.content, content)
         self.assertTrue(user_message.flags.mentioned)
 
+    def send_and_verify_wildcard_mention_message(self,
+                                                 sender_name: str,
+                                                 test_fails: bool=False,
+                                                 sub_count: int=16) -> None:
+        sender = self.example_user(sender_name)
+        content = "@**all** test wildcard mention"
+        with mock.patch(
+            "zerver.lib.message.num_subscribers_for_stream_id", return_value=sub_count
+        ):
+            if not test_fails:
+                msg_id = self.send_stream_message(sender, "test_stream", content)
+                result = self.api_get(sender, '/json/messages/' + str(msg_id))
+                self.assert_json_success(result)
+
+            else:
+                with self.assertRaisesRegex(JsonableError,
+                                            "You do not have permission to use wildcard mentions in this stream."):
+                    self.send_stream_message(sender, "test_stream", content)
+
+    def test_wildcard_mention_restrictions(self) -> None:
+        cordelia = self.example_user("cordelia")
+        iago = self.example_user("iago")
+        polonius = self.example_user("polonius")
+        realm = cordelia.realm
+
+        stream_name = "test_stream"
+        self.subscribe(cordelia, stream_name)
+        self.subscribe(iago, stream_name)
+        self.subscribe(polonius, stream_name)
+
+        do_set_realm_property(
+            realm, "wildcard_mention_policy", Realm.WILDCARD_MENTION_POLICY_EVERYONE
+        )
+        self.send_and_verify_wildcard_mention_message("polonius")
+
+        do_set_realm_property(
+            realm, "wildcard_mention_policy", Realm.WILDCARD_MENTION_POLICY_MEMBERS
+        )
+        self.send_and_verify_wildcard_mention_message("polonius", test_fails=True)
+        # There is no restriction on small streams.
+        self.send_and_verify_wildcard_mention_message("polonius", sub_count=10)
+        self.send_and_verify_wildcard_mention_message("cordelia")
+
+        do_set_realm_property(
+            realm,
+            "wildcard_mention_policy",
+            Realm.WILDCARD_MENTION_POLICY_FULL_MEMBERS
+        )
+        do_set_realm_property(realm, "waiting_period_threshold", 10)
+        iago.date_joined = timezone_now()
+        cordelia.date_joined = timezone_now()
```

Comment:
We need to save iago/cordelia here; otherwise, this test will start failing on any local database where the users are more than 10 days old :)